### PR TITLE
Issue #5: Script to iterate over git history and apply certain config  to code.

### PR DESCRIPTION
**Pull request by HuGanghui**
Issue #5: Script to iterate over git history and apply certain config  to code.

using generatePatch() can get patch file between any two commits in one repo. 
after getting the patch file. We can using it to test patch-filter, I make a draft PR(https://github.com/checkstyle/checkstyle/pull/8318) that put patch-filter into checkstyle to do checkstyle-tester.

three report examples on guava:

report file: https://huganghui.github.io/generatePatchReport/DiffReport/

baseBranch: master (the difference with patchBranch is not have new feature patch-filter's code), baseConfig with severity ignore. so no violations in report.

baseConfig:
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="ignore"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>

```

patchConfig:
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionPatchFilter">
        <property name="file" value="/Users/hgh/Documents/GitHub/patch-filters/DiffReport/patch.txt"/>
    </module>

    <!--<module name="Header">-->
        <!--<property name="headerFile" value="${checkstyle.header.file}"/>-->
        <!--<property name="fileExtensions" value="java"/>-->
        <!--<property name="id" value="header"/>-->
    <!--</module>-->
    <!--<module name="RegexpHeader">-->
        <!--<property name="headerFile" value="${checkstyle.regexp.header.file}"/>-->
        <!--<property name="fileExtensions" value="java"/>-->
    <!--</module>-->

    <!-- Javadoc Comments -->
    <module name="JavadocPackage">
        <property name="allowLegacy" value="false"/>
    </module>

    <!-- Miscellaneous -->
    <module name="NewlineAtEndOfFile"/>
    <module name="Translation">
        <property name="requiredTranslations" value="de, fr, fi, es, pt, ja, tr, zh"/>
    </module>
    <module name="UniqueProperties"/>
    <module name="OrderedProperties" />

    <!-- Regexp -->
    <module name="RegexpMultiline"/>
    <module name="RegexpMultiline">
        <property name="format" value="\r?\n[\t ]*\r?\n[\t ]*\r?\n"/>
        <property name="fileExtensions" value="java,xml,properties"/>
        <property name="message" value="Unnecessary consecutive lines"/>
    </module>
    <module name="RegexpMultiline">
        <property name="format" value="/\*\*\W+\* +\p{javaLowerCase}"/>
        <property name="fileExtensions" value="java"/>
        <property name="message"
                  value="First sentence in a comment should start with a capital letter"/>
    </module>
    <module name="RegexpMultiline">
        <property name="format" value="^\s*$" />
        <property name="matchAcrossLines" value="true" />
        <property name="message" value="Empty file is not allowed" />
    </module>
    <module name="RegexpSingleline">
        <property name="format" value="\s+$"/>
        <property name="minimum" value="0"/>
        <property name="maximum" value="0"/>
    </module>
    <module name="RegexpSingleline">
        <property name="format" value="/\*\* +\p{javaLowerCase}"/>
        <property name="fileExtensions" value="java"/>
        <property name="message"
                  value="First sentence in a comment should start with a capital letter"/>
    </module>
    <module name="RegexpSingleline">
        <property name="format" value="^(?!(.*http|import)).{101,}$"/>
        <property name="fileExtensions" value="g, g4"/>
        <property name="message" value="Line should not be longer than 100 symbols"/>
    </module>
    <module name="RegexpSingleline">
        <property name="format"
                  value="^(?!(\s*,?\s*&lt;a href=&quot;[^&quot;]+&quot;&gt;|.*http)).{101,}$"/>
        <property name="fileExtensions" value="xml, vm"/>
        <property name="message" value="Line should not be longer than 100 symbols"/>
    </module>
    <!--
     Links to .dtd files should start with "/", "http://" or "https://",
     otherwise they will be broken after archiving the documentation.
     See https://github.com/checkstyle/checkstyle/issues/7340 for details.
    -->
    <module name="RegexpSingleline">
        <property name="format" value="href=&quot;(?!\/|https?:\/\/).*?\.dtd&quot;"/>
        <property name="fileExtensions" value="xml, vm"/>
        <property name="message"
                  value="Relative links to DTD files are prohibited. Please use absolute path or uri instead."/>
    </module>
    <module name="RegexpSingleline">
        <property name="id" value="noSourceforgeNetLinks"/>
        <property name="format" value="checkstyle\.sourceforge\.net"/>
        <property name="message"
                  value="Old site links should not be used, please use https://checkstyle.org"/>
    </module>
    <module name="RegexpSingleline">
        <property name="id" value="noSourceforgeIoLinks"/>
        <property name="format" value="checkstyle\.sourceforge\.io"/>
        <property name="message"
                  value="Old site links should not be used, please use https://checkstyle.org"/>
    </module>
    <module name="RegexpSingleline">
        <property name="id" value="noPackageCommentWithOtherVisibility"/>
        <property name="format" value="/\*\s+package\s+\*/\s+(private|protected|public)"/>
        <property name="fileExtensions" value="java"/>
        <property name="message"
                  value="Package comment marker should not be used if other visibility is defined"/>
    </module>
    <module name="RegexpOnFilename" />
    <module name="RegexpOnFilename">
        <property name="folderPattern" value="[\\/]src[\\/]\w+[\\/]java[\\/]"/>
        <property name="fileNamePattern" value="\.java$"/>
        <property name="match" value="false"/>
        <message key="regexp.filepath.mismatch"
                 value="Only java files should be located in the ''src/*/java'' folders."/>
    </module>
    <module name="RegexpOnFilename">
        <property name="folderPattern" value="[\\/]src[\\/]xdocs[\\/]"/>
        <property name="fileNamePattern" value="\.(xml)|(vm)$"/>
        <property name="match" value="false"/>
        <message key="regexp.filepath.mismatch"
                 value="All files in the ''src/xdocs'' folder should have the ''xml'' or ''vm'' extension."/>
    </module>
    <module name="RegexpOnFilename">
        <property name="folderPattern" value="[\\/]src[\\/]it[\\/]java[\\/]"/>
        <property name="fileNamePattern" value="^((\w+Test)|(\w+TestSupport)|(Abstract\w+))\.java$"/>
        <property name="match" value="false"/>
        <message key="regexp.filepath.mismatch"
                 value="All files in the ''src/it/java'' folder
                      should be named ''*Test.java'' or ''Abstract*.java''."/>
    </module>

    <!-- Size Violations -->
    <module name="FileLength">
        <property name="fileExtensions" value="java"/>
    </module>
    <module name="LineLength">
        <property name="fileExtensions" value="java"/>
        <property name="max" value="100"/>
        <property name="ignorePattern" value="^ *\* *([^ ]+|\{@code .*|&lt;a href=&quot;[^&quot;]+&quot;&gt;)$"/>
    </module>

    <!-- Whitespace -->
    <module name="FileTabCharacter">
        <property name="eachLine" value="false"/>
    </module>

    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>
```

patchBranch: patch-filter, patchConfig with patch-filter, violations are filtered by patch-filter, so also no violations in report.

### CODE REVIEW

**Comment by romani** [comment]
Is it executable ? Does it produce any result ? Please share report as soon as it produce result

**Comment by HuGanghui** [comment]
Is this version of script ok? Or we need some improvements, please give me some advice.

**Comment by romani** [comment]
I do not understand this report. I do not understand why you need diff report at all.

All you need is
1) to mvn install this project to local repo.
2) update checkstyle-tester to use new jar in dependency of checkstyle in pom.xml of checkstyle-tester folder.
3) use single config.

4) execute checkstyle-tester as it is done usually, but with generation of patch file for filter.
5) collect violations and print them to special !!!! report, you can not reuse existing. You should have folder for each commit with: patch, list violations,    files that are with violations (before and after the patch)

**Comment by romani** [comment]
We can reuse diff report.

But we should execute on two different configs: with severity ignore in one config and second config with filter. Diff between two execution is clear effect of filter. 
But we need to execute full run on each commit, and each commit result in full diff report in special folder. 
Report from such regression tool is list of folfers (separate report for each commit)

**Comment by HuGanghui** [comment]
> with severity ignore in one config
What it mean, I do not fully undersand
> But we need to execute full run on each commit, and each commit result in full diff report in       > special folder.
OK, I will do it, old version is to make sure that it can run.
> In each diff report (for each commit) should be link to patch file to review patch online easily
OK, I will try it.

**Comment by rdiachenko** [comment]
@HuGanghui 
> What it mean, I do not fully undersand

First execution with a patch filter turned OFF, you can use `<property name="severity" value="ignore"/>` inside the filter to do that. Second execution with a patch filter turned ON, you may remove or comment out that `severity` property inside the filter to do that.

**Comment by romani** [comment]
I meant:
https://github.com/checkstyle/contribution/blob/99ad2d9de0ddc4da563fdddd80aab6a7dad3bfa7/checkstyle-tester/checks-nonjavadoc-error.xml#L17
We need to compare state of no violations with violations by config with filter. We emulate situation when checkstyle was not in project (no violations at all) and checkstyle is introduced (Checks + patch filter)

or we can completely skip such execution and just generate empty xml report file to let https://github.com/checkstyle/contribution/tree/master/patch-diff-report-tool think there was no violations.

File: src/main/java/com/github/checkstyle/generatepatchfile/GeneratePatchFile.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            System.out.println("LogCommit: " + commit +
+                    ", name: " + commit.getName() + ", id: " + commit.getId().getName());
+            count++;
+            commitList.add(commit);
+        }
+        System.out.println(count);
+        return commitList;
+    }
+
+    private void generateTwoCommitDiffPatch(RevCommit commitOld, RevCommit commitNew) throws Exception {
```
Comment (romani) [inline src/main/java/com/github/checkstyle/generatepatchfile/GeneratePatchFile.java:126]: please stop any activity to re-implement `git format-patch --stdout HEAD~2`  we will test only git and only on linux and we do not need java implementation of it during POC, probably will not need it ever.
You need to just execute shell command `git format-patch --stdout HEAD~1` , `git format-patch --stdout HEAD~2` ,..... to get patch file. To generate patch for two commits in one file `git format-patch -2 --stdout`

Please keep this implementation as is and make implementations for usage of shell execution of git commands.

File: src/main/java/com/github/checkstyle/generatepatchfile/GeneratePatchFile.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    private void checkout(String commitName) throws Exception {
+        git.checkout().setName(commitName).call();
+    }
+
+    private File generate()
+            throws InterruptedException, IOException {
+        final Process process = new ProcessBuilder()
+                .directory(new File(testerPath))
+                .command(
+                        "groovy", "diff.groovy",
```
Comment (romani) [inline src/main/java/com/github/checkstyle/generatepatchfile/GeneratePatchFile.java:172]: you do not need diff.groovy

please use inner generator of diff report based on 2 checkstyle xml reports https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/diff.groovy#L209

File: src/main/resources/baseConfig.xml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
+    <property name="haltOnException" value="false"/>
+
+    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
+    <module name="BeforeExecutionExclusionFileFilter">
+        <property name="fileNamePattern" value="module\-info\.java$" />
+    </module>
+
+    <module name="TreeWalker">
```
Comment (romani) [inline src/main/resources/baseConfig.xml:20]: remove TreeWalker at all, we are not ready for this validaation.

please all these Checks:
https://github.com/checkstyle/checkstyle/blob/ec4d06712ab203d31d73c5c6d5c46067f3a6d5b3/config/checkstyle_checks.xml#L60-L190
you can skip Translation Check and checks that reuqires extra config file (Header and RegExpHeader)

Comment (HuGanghui) [reply src/main/resources/baseConfig.xml]: done

File: src/main/java/com/github/checkstyle/generatepatchfile/GeneratePatchFile.java (hunk 2 above)
Comment (HuGanghui) [reply src/main/java/com/github/checkstyle/generatepatchfile/GeneratePatchFile.java:178]: done, understand

File: src/main/resources/patchConfig.xml [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
+    <property name="haltOnException" value="false"/>
+
+    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
+    <module name="BeforeExecutionExclusionFileFilter">
+        <property name="fileNamePattern" value="module\-info\.java$" />
+    </module>
+
+    <module name="com.puppycrawl.tools.checkstyle.filters.SuppressionPatchFilter">
+        <property name="file" value="/Users/hgh/Documents/GitHub/patch-filters/DiffReport/patch.txt"/>
```
Comment (baratali) [inline src/main/resources/patchConfig.xml:21]: Please change the path a to a patch file. Others don't have this folder.
Also you generate patch.txt files in current dir https://github.com/checkstyle/patch-filters/pull/17/files#diff-bddb2f25480aacedb24f2f1035abc85dR143, you should use the same here

Comment (HuGanghui) [reply src/main/resources/patchConfig.xml:21]: so is necessary to using org.dom4j or other library to dynamic add this path or we have some simple way to modify it?

Comment (baratali) [reply src/main/resources/patchConfig.xml:21]: I guess `./DiffReport/patch.txt` will work. Can you please test it?

Comment (HuGanghui) [reply src/main/resources/patchConfig.xml:21]: I test it, and it is not ok, and I understand the reason is that this patchConfig is used by checkstyle-tester and not by this project itself, so absolute path is necessary

**Pull request merged**